clickhouse-core/src/main/scala/xenon/clickhouse/hash/HashFunc.scala

```diff
@@ -0,0 +1,29 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package xenon.clickhouse.hash
+
+import scala.reflect.ClassTag
+
+abstract class HashFunc[T: ClassTag] {
+  def applyHash(input: Array[Byte]): T
+  def combineHashes(h1: T, h2: T): T
+
+  final def executeAny(input: Any): T =
```

Comment:
seems not necessary, as we only allow UTF-8 bytes now.

I suppose the following method is enough now.

```
def apply(input: Array[Byte]): Long
def apply(input: Array[Array[Byte]]): Long = input.map(apply).reduce(combineHashes)
```

Reply:
I am copying the similar logic from [Clickhouse](https://github.com/ClickHouse/ClickHouse/blob/v23.5.3.24-stable/src/Functions/FunctionsHashing.h#L1315), so if we want to support more types, we can easily add type -> bytearray handling here in `executeAny`.

Reply:
I see your point, one drawback is that Java/Scala generic/template system is not efficient as cpp does on primitive types.

I don't have a strong opinion now. Let's focus on the correctness and functionality first, and do investigations on performance later.